src/devices/SoftPwm/SoftwarePwmChannel.cs

```diff
@@ -105,106 +110,89 @@ public SoftwarePwmChannel(int pinNumber, int frequency = 400, double dutyCycle =
 
-            UpdatePulseWidthParameters();
+            if (usePrecisionTimer)
+            {
+                _stopwatch = new Stopwatch();
+                _stopwatch.Start();
+                _thread.Priority = ThreadPriority.Highest;
+            }
+
+            _controller.OpenPin(_pin, PinMode.Output);
 
-            _runningThread = new Thread(RunSoftPWM);
-            _runningThread.Start();
+            _thread.Start();
         }
 
         private void UpdatePulseWidthParameters()
         {
-            _periodInMilliseconds = 1000.0 / _frequency;
-            _pulseWidthInMilliseconds = _dutyCycle * _periodInMilliseconds;
+            double cycleTicks = TimeSpan.TicksPerSecond / (double)_frequency;
+
+            double pinHighTicks = cycleTicks * _dutyCycle;
+            _pinHighTime = TimeSpan.FromTicks((long)pinHighTicks);
+
+            double pinLowTicks = cycleTicks - pinHighTicks;
+            _pinLowTime = TimeSpan.FromTicks((long)pinLowTicks);
         }
 
-        private void RunSoftPWM()
+        private void Run()
         {
-            if (_usePrecisionTimer)
+            while (!_isTerminating)
             {
-                Thread.CurrentThread.Priority = ThreadPriority.Highest;
-            }
-
-            while (_runThread)
-            {
-                // Write the pin high for the appropriate length of time
-                if (_isRunning)
+                if (!_isRunning)
                 {
-                    if (_pulseWidthInMilliseconds != 0)
-                    {
-                        _controller.Write(_pin, PinValue.High);
-                        _isStopped = false;
-                    }
-
-                    // Use the wait helper method to wait for the length of the pulse
-                    if (_usePrecisionTimer)
-                    {
-                        Wait(_pulseWidthInMilliseconds);
-                    }
-                    else
-                    {
-                        Task.Delay(TimeSpan.FromMilliseconds(_pulseWidthInMilliseconds)).Wait();
-                    }
-
-                    // The pulse if over and so set the pin to low and then wait until it's time for the next pulse
-                    _controller.Write(_pin, PinValue.Low);
+                    Thread.Yield();
+                    continue;
+                }
 
-                    if (_usePrecisionTimer)
-                    {
-                        Wait(_periodInMilliseconds - _pulseWidthInMilliseconds);
-                    }
-                    else
-                    {
-                        Task.Delay(TimeSpan.FromMilliseconds(_periodInMilliseconds - _pulseWidthInMilliseconds)).Wait();
-                    }
+                if (_pinHighTime != TimeSpan.Zero)
+                {
+                    _controller.Write(_pin, PinValue.High);
+                    Wait(_pinHighTime);
                 }
-                else
+
+                if (_pinLowTime != TimeSpan.Zero)
```

Comment:
The old logic was checking if the high time was zero, but not if the low time was. We do both.